File: rec_llm_python/app/watcher/folder_watcher.py
"""RecLLM Python Core — Folder Watcher (watchdog-based)"""

import logging
import time
from pathlib import Path
from dataclasses import dataclass, field
from threading import Thread

from app.config import AUDIO_EXTENSIONS

logger = logging.getLogger(__name__)


@dataclass
class WatcherState:
    active: bool = False
    folder_path: str | None = None
    known_files: set = field(default_factory=set)
    on_new_files: list = field(default_factory=list)  # callbacks


class FolderWatcher:
    """Watch a folder for new audio files and notify callbacks."""

    def __init__(self):
        self._state = WatcherState()
        self._thread: Thread | None = None
        self._stop_flag = False

    @property
    def active(self) -> bool:
        return self._state.active

    @property
    def folder_path(self) -> str | None:
        return self._state.folder_path

    @property
    def known_file_count(self) -> int:
        return len(self._state.known_files)

    def on_new_files(self, callback):
        """Register a callback for new file detection."""
        self._state.on_new_files.append(callback)

    def start(self, folder_path: str) -> dict:
        """Start watching a folder."""
        path = Path(folder_path)
        if not path.exists() or not path.is_dir():
            return {"ok": False, "error": f"Folder not found: {folder_path}"}

        if self._state.active:
            self.stop()

        # Scan existing files
        existing = self._scan_folder(path)
        self._state.known_files = existing
        self._state.folder_path = folder_path
        self._state.active = True
        self._stop_flag = False

        # Start polling thread
        self._thread = Thread(target=self._poll_loop, daemon=True)
        self._thread.start()

        logger.info("Folder watcher started: %s (%d existing files)", folder_path, len(existing))
        return {"ok": True, "fileCount": len(existing)}

    def stop(self):
        """Stop watching."""
        self._stop_flag = True
        self._state.active = False
        self._state.folder_path = None
        self._state.known_files.clear()
        if self._thread:
            self._thread.join(timeout=5)
            self._thread = None
        logger.info("Folder watcher stopped")

    def status(self) -> dict:
        return {
            "active": self._state.active,
            "folderPath": self._state.folder_path,
            "knownFileCount": len(self._state.known_files),
        }

    def _poll_loop(self):
        """Poll for new files every 2 seconds."""
        while not self._stop_flag and self._state.active:
            try:
                if self._state.folder_path:
                    path = Path(self._state.folder_path)
                    current = self._scan_folder(path)
                    new_files = current - self._state.known_files

                    if new_files:
                        self._state.known_files = current
                        self._notify_new_files(new_files)
            except Exception as e:
                logger.warning("Folder watcher error: %s", e)

            time.sleep(2.0)

    def _scan_folder(self, folder: Path) -> set[str]:
        """Scan folder for audio files (non-recursive for safety)."""
        files = set()
        try:
            for f in folder.iterdir():
                if f.is_file() and f.suffix.lstrip(".").lower() in AUDIO_EXTENSIONS:
                    files.add(str(f))
        except PermissionError:
            logger.warning("Permission denied: %s", folder)
        return files

    def _notify_new_files(self, new_files: set[str]):
        """Notify callbacks about new files."""
        file_metas = []
        for fp in new_files:
            p = Path(fp)
            file_metas.append({
                "fileName": p.name,
                "filePath": str(p),
                "sizeBytes": p.stat().st_size if p.exists() else 0,
                "extension": p.suffix.lstrip(".").lower(),
            })

        logger.info("Folder watcher: %d new files detected", len(file_metas))
        for callback in self._state.on_new_files:
            try:
                callback(file_metas)
            except Exception as e:
                logger.warning("Watcher callback error: %s", e)

File: electron/folder-watcher.ts
/**
 * Folder Watcher Service
 *
 * Watches a configured folder for new audio files and emits events
 * when new files are detected. Uses Node.js fs.watch (no external deps).
 */

import { ipcMain, app, BrowserWindow } from 'electron';
import fs from 'fs';
import path from 'path';
import { writeLog } from './history';

const AUDIO_EXTENSIONS = new Set(['.mp3', '.wav', '.m4a', '.mp4', '.aac', '.flac', '.ogg', '.wma']);
const POLL_INTERVAL_MS = 5000; // Check every 5 seconds
const DEBOUNCE_MS = 2000; // Wait 2s after file appears (ensure write is complete)

interface WatcherState {
  active: boolean;
  folderPath: string | null;
  knownFiles: Set<string>;
  pollTimer: NodeJS.Timeout | null;
}

const state: WatcherState = {
  active: false,
  folderPath: null,
  knownFiles: new Set(),
  pollTimer: null,
};

function isAudioFile(filePath: string): boolean {
  const ext = path.extname(filePath).toLowerCase();
  return AUDIO_EXTENSIONS.has(ext);
}

function scanFolder(folderPath: string): string[] {
  const results: string[] = [];
  try {
    const entries = fs.readdirSync(folderPath, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(folderPath, entry.name);
      if (entry.isDirectory()) {
        // Recursive scan
        results.push(...scanFolder(fullPath));
      } else if (entry.isFile() && isAudioFile(entry.name)) {
        results.push(fullPath);
      }
    }
  } catch {
    // Folder may have been deleted or permissions changed
  }
  return results;
}

function getFileStableSize(filePath: string): number {
  try {
    return fs.statSync(filePath).size;
  } catch {
    return -1;
  }
}

async function checkForNewFiles(): Promise<void> {
  if (!state.active || !state.folderPath) return;

  const currentFiles = scanFolder(state.folderPath);
  const newFiles: string[] = [];

  for (const file of currentFiles) {
    if (!state.knownFiles.has(file)) {
      // Verify file is not still being written (size stable)
      const size1 = getFileStableSize(file);
      if (size1 <= 0) continue;

      // Wait briefly and check again
      await new Promise((r) => setTimeout(r, DEBOUNCE_MS));
      const size2 = getFileStableSize(file);

      if (size2 === size1 && size2 > 0) {
        // File is stable — it's a new complete file
        newFiles.push(file);
        state.knownFiles.add(file);
      }
      // If sizes differ, file is still being written — skip this cycle, catch next time
    }
  }

  // Remove files that no longer exist from known set
  for (const known of state.knownFiles) {
    if (!fs.existsSync(known)) {
      state.knownFiles.delete(known);
    }
  }

  // Notify renderer of new files
  if (newFiles.length > 0) {
    const win = BrowserWindow.getAllWindows()[0];
    if (win) {
      const fileMetas = newFiles.map((filePath) => {
        const stat = fs.statSync(filePath);
        const ext = path.extname(filePath).slice(1).toLowerCase();
        return {
          id: `watch_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
          fileName: path.basename(filePath),
          filePath,
          sizeBytes: stat.size,
          extension: ext,
        };
      });
      win.webContents.send('watcher:newFiles', fileMetas);
      await writeLog('INFO', `Folder watcher detected ${newFiles.length} new file(s)`, newFiles.map((f) => path.basename(f)).join(', '));
    }
  }
}

function startPolling(): void {
  if (state.pollTimer) return;
  state.pollTimer = setInterval(() => {
    checkForNewFiles().catch(() => {});
  }, POLL_INTERVAL_MS);
}

function stopPolling(): void {
  if (state.pollTimer) {
    clearInterval(state.pollTimer);
    state.pollTimer = null;
  }
}

export function registerFolderWatcherHandlers(): void {
  ipcMain.handle('watcher:start', async (_event, folderPath: unknown): Promise<{ ok: boolean; error?: string; fileCount?: number }> => {
    if (typeof folderPath !== 'string' || !folderPath) {
      return { ok: false, error: 'Invalid folder path' };
    }
    if (!fs.existsSync(folderPath) || !fs.statSync(folderPath).isDirectory()) {
      return { ok: false, error: 'Folder does not exist' };
    }

    // Initialize known files (don't trigger for existing files)
    const existingFiles = scanFolder(folderPath);
    state.knownFiles = new Set(existingFiles);
    state.folderPath = folderPath;
    state.active = true;

    startPolling();
    await writeLog('INFO', `Folder watcher started`, `path=${folderPath} existing=${existingFiles.length}`);

    return { ok: true, fileCount: existingFiles.length };
  });

  ipcMain.handle('watcher:stop', async (): Promise<{ ok: boolean }> => {
    state.active = false;
    state.folderPath = null;
    state.knownFiles.clear();
    stopPolling();
    await writeLog('INFO', 'Folder watcher stopped');
    return { ok: true };
  });

  ipcMain.handle('watcher:status', async (): Promise<{ active: boolean; folderPath: string | null; knownFileCount: number }> => {
    return {
      active: state.active,
      folderPath: state.folderPath,
      knownFileCount: state.knownFiles.size,
    };
  });
}
